### PR TITLE
[WIP] Rust course project 2 tests and example solution 

### PR DESCRIPTION
**Pull request by sticnarf**

### CODE REVIEW

File: rust/projects/file-io/Cargo.toml [hunk 1]
```diff
@@ -7,4 +7,6 @@ edition = "2018"
 
 [dependencies]
 clap = "2.32.0"
-failure = "0.1.5"
\ No newline at end of file
+failure = "0.1.5"
+serde = { version = "1.0.89", features = ["derive"] }
```
Comment (brson) [inline rust/projects/file-io/Cargo.toml]: I did not know you could get the derive macros directly from the serde crate. Neat!

File: rust/projects/file-io/src/kv.rs [hunk 2]
```diff
@@ -40,4 +66,28 @@ impl KvStore {
...
     }
+
+    fn load_from_wal(wal_path: impl AsRef<Path>) -> Result<HashMap<String, String>> {
+        let mut map = HashMap::new();
+        let reader = BufReader::new(File::open(wal_path)?);
+        let stream = Deserializer::from_reader(reader).into_iter::<SetCommand>();
+        for set_cmd in stream {
+            let set_cmd = set_cmd?;
+            map.insert(set_cmd.key, set_cmd.value);
+        }
```
Comment (brson) [inline rust/projects/file-io/src/kv.rs]: I'm glad this code was so easy. I was afraid serde wouldn't deserialize multiple records in sequence like that.

File: rust/projects/file-io/tests/tests.rs [hunk 3]
```diff
@@ -113,12 +142,29 @@ fn binary_path() -> PathBuf {
     path
 }
 
+fn temp_dir(name: &str) -> PathBuf {
+    let mut temp_dir = env::temp_dir();
+    temp_dir.push(name);
+    create_dir_all(&temp_dir).expect("unable to create working directory");
+    temp_dir
+}
```
Comment (brson) [inline rust/projects/file-io/tests/tests.rs]: From the looks of this the temp dirs are not deleted up after the test. Can you instead use the tempfile crate's tempdir, which has destructors that automatically clean up.

File: rust/projects/file-io/src/kv.rs [hunk 4]
```diff
@@ -15,22 +20,43 @@ use crate::Result;
...
+        })
     }
 
     /// Sets the value of a string key to a string.
     ///
     /// If the key already exists, the previous value will be overwritten.
     pub fn set(&mut self, key: String, value: String) -> Result<()> {
-        self.map.insert(key, value);
+        let cmd = SetCommand::new(key, value);
+        to_writer(&mut self.wal_writer, &cmd)?;
```
Comment (brson) [inline rust/projects/file-io/src/kv.rs]: It took me a few seconds to figure out that `to_writer` isn't in this crate, but from serde_json.

It's common in Rust to import types, but to call functions through their module name, like `serde_json::to_writer`. Will you do that here for clarity?

File: rust/projects/file-io/tests/tests.rs (hunk 3 above)
Comment (sticnarf) [reply rust/projects/file-io/tests/tests.rs]: I was avoiding introducing dependencies for tests. If we do so, we'll either change to use workspaces or just add dev-dependencies to Cargo.toml (maybe mention that in project 1).

**Comment by sticnarf** [comment]
@brson After some investigation, I confirm serde_json writes in small pieces, so a `BufWriter` counts. 

You're right. To ensure we do not lost data in the buffer, we need a flush on every write of the user.

File: rust/projects/file-io/tests/tests.rs (hunk 3 above)
Comment (sticnarf) [reply rust/projects/file-io/tests/tests.rs]: This way, students copy the content of Cargo.toml at first, and copy the tests of each project, then everything will go fine.

**Comment by brson** [comment]
Thanks @sticnarf. I went ahead and merged. Please continue the project in a new branch.

**Pull request merged**